File: ironpythoninaction/chapter15/15.3/PluginEmbeddingVB/PluginEmbeddingVB/Engine.vb
Imports System.IO
Imports System.Reflection
Imports System.Text

Imports IronPython.Hosting
Imports IronPython.Runtime
Imports IronPython.Runtime.Exceptions
Imports Microsoft.Scripting
Imports Microsoft.Scripting.Runtime
Imports Microsoft.Scripting.Hosting

Imports Plugins

Friend Class PythonStream
    Inherits MemoryStream

    Dim _output As TextBox

    Public Sub New(ByVal textbox As TextBox)
        _output = textbox
    End Sub

    Public Overrides Sub Write(ByVal buffer As Byte(), ByVal offset As Integer, ByVal count As Integer)
        _output.AppendText(Encoding.UTF8.GetString(buffer, offset, count))
    End Sub
End Class

Friend Class Engine
    Dim _engine As ScriptEngine
    Dim _runtime As ScriptRuntime
    Dim _box As TextBox

    Public ReadOnly Property Plugins() As List(Of PluginBase)
        Get
            Return PluginStore.Plugins
        End Get
    End Property

    Public Sub New(ByVal textbox As TextBox)
        _engine = Python.CreateEngine()
        _runtime = _engine.Runtime

        _box = textbox

        SetStreams()
        Dim rootDir As String = AddAssemblies()
        LoadPlugins(rootDir)
    End Sub

    Public Sub SetStreams()
        Dim stream As PythonStream = New PythonStream(_box)
        _runtime.IO.SetOutput(stream, Encoding.UTF8)
        _runtime.IO.SetErrorOutput(stream, Encoding.UTF8)
    End Sub

    Public Function AddAssemblies() As String
        Dim mainAssembly As Assembly = Assembly.GetExecutingAssembly()

        Dim rootDir As String = Directory.GetParent(mainAssembly.Location).FullName
        Dim pluginsPath As String = Path.Combine(rootDir, "Plugins.dll")

        Dim pluginsAssembly As Assembly = Assembly.LoadFile(pluginsPath)

        _runtime.LoadAssembly(mainAssembly)
        _runtime.LoadAssembly(pluginsAssembly)
        _runtime.LoadAssembly(GetType(String).Assembly)
        _runtime.LoadAssembly(GetType(Uri).Assembly)

        Return rootDir
    End Function

    Public Sub LoadPlugins(ByVal rootDir As String)
        Dim pluginsDir As String = Path.Combine(rootDir, "plugins")
        For Each path As String In Directory.GetFiles(pluginsDir)
            If path.ToLower().EndsWith(".py") Then
                CreatePlugin(path)
            End If
        Next
    End Sub

    Public Sub CreatePlugin(ByVal path As String)
        Try
            Dim script As ScriptSource = _engine.CreateScriptSourceFromFile(path)
            Dim code As CompiledCode = script.Compile()
            Dim scope As ScriptScope = _engine.CreateScope()
            script.Execute(scope)
        Catch e As SyntaxErrorException
            Dim msg As String = "Syntax error in ""{0}"""
            ShowError(msg, System.IO.Path.GetFileName(path), e)
        Catch e As SystemExitException
            Dim msg As String = "SystemExit in ""{0}"""
            ShowError(msg, System.IO.Path.GetFileName(path), e)
        Catch e As Exception
            Dim msg As String = "Error loading plugin ""{0}"""
            ShowError(msg, System.IO.Path.GetFileName(path), e)
        End Try
    End Sub

    Public Sub ShowError(ByVal title As String, ByVal name As String, ByVal e As Exception)
        Dim caption As String = String.Format(title, name)
        Dim eo As ExceptionOperations = _engine.GetService(Of ExceptionOperations)()
        Dim traceback As String = eo.FormatException(e)
        MessageBox.Show(traceback, caption, MessageBoxButtons.OK, MessageBoxIcon.Error)
    End Sub

    Public Sub ExecutePluginAtIndex(ByVal index As Integer)
        Dim plugin As PluginBase = Plugins(index)

        Try
            plugin.Execute(_box)
        Catch e As Exception
            Dim msg As String = "Error executing plugin ""{0}"""
            ShowError(msg, plugin.Name, e)
        End Try
    End Sub
End Class

File: ironpythoninaction/chapter15/15.3/EmbeddingPlugin/EmbeddingPlugin/Engine.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Text;
using System.Windows.Forms;

using IronPython.Hosting;
using IronPython.Runtime;
using IronPython.Runtime.Exceptions;
using Microsoft.Scripting;
using Microsoft.Scripting.Runtime;
using Microsoft.Scripting.Hosting;

using Plugins;


namespace EmbeddingPlugin
{
    internal class PythonStream: MemoryStream
    {
        TextBox _output;
        public PythonStream(TextBox textbox)
        {
            _output = textbox;
        }

        public override void Write(byte[] buffer, int offset, int count)
        {
            _output.AppendText(Encoding.UTF8.GetString(buffer, offset, count));
        }
    }

    internal class Engine
    {
        ScriptEngine _engine;
        ScriptRuntime _runtime;
        TextBox _box;

        public List<PluginBase> Plugins
        {
            get { return PluginStore.Plugins; }
        }


        public Engine(TextBox textbox)
        {
            _engine = Python.CreateEngine();
            _runtime = _engine.Runtime;
            _box = textbox;

            SetStreams();
            string rootDir = AddAssemblies();
            LoadPlugins(rootDir);
        }

        public void SetStreams()
        {
            PythonStream stream = new PythonStream(_box);
            _runtime.IO.SetOutput(stream, Encoding.UTF8);
            _runtime.IO.SetErrorOutput(stream, Encoding.UTF8);
        }

        public string AddAssemblies()
        {
            Assembly mainAssembly = Assembly.GetExecutingAssembly();

            string rootDir = Directory.GetParent(mainAssembly.Location).FullName;
            string pluginsPath = Path.Combine(rootDir, "Plugins.dll");

            Assembly pluginsAssembly = Assembly.LoadFile(pluginsPath);

            _runtime.LoadAssembly(mainAssembly);
            _runtime.LoadAssembly(pluginsAssembly);
            _runtime.LoadAssembly(typeof(String).Assembly);
            _runtime.LoadAssembly(typeof(Uri).Assembly);

            return rootDir;
        }

        public void LoadPlugins(string rootDir)
        {
            string pluginsDir = Path.Combine(rootDir, "plugins");
            foreach (string path in Directory.GetFiles(pluginsDir))
            {
                if (path.ToLower().EndsWith(".py"))
                {
                    CreatePlugin(path);
                }
            }
        }

        public void CreatePlugin(string path)
        {
            try
            {
                ScriptSource script = _engine.CreateScriptSourceFromFile(path);
                CompiledCode code = script.Compile();
                ScriptScope scope = _engine.CreateScope();
                script.Execute(scope);
            }
            catch (SyntaxErrorException e)
            {
                string msg = "Syntax error in \"{0}\"";
                ShowError(msg, Path.GetFileName(path), e);
            }
            catch (SystemExitException e)
            {
                string msg = "SystemExit in \"{0}\"";
                ShowError(msg, Path.GetFileName(path), e);
            }

            catch (Exception e)
            {
                string msg = "Error loading plugin \"{0}\"";
                ShowError(msg, Path.GetFileName(path), e);
            }
        }

        public void ShowError(string title, string name, Exception e)
        {
            string caption = String.Format(title, name);
            ExceptionOperations eo = _engine.GetService<ExceptionOperations>();
            string error = eo.FormatException(e);
            MessageBox.Show(error, caption, MessageBoxButtons.OK, MessageBoxIcon.Error);

        }

        public void ExecutePluginAtIndex(int index)
        {
            PluginBase plugin = Plugins[index];

            try
            {
                plugin.Execute(_box);
            }
            catch (Exception e)
            {
                string msg = "Error executing plugin \"{0}\"";
                ShowError(msg, plugin.Name, e);
            }
        }
    }
}
